api/envoy/api/v2/cds.proto

```diff
@@ -236,6 +236,65 @@ message Cluster {
   //   verification.
   auth.UpstreamTlsContext tls_context = 11;
 
+  // TransportSocketMatch specifies what transport socket config will be used
+  // when the matches conditions are satisfied.
+  message TransportSocketMatch {
+    // The name of the match.
+    string name = 1;
+
+    // Optional endpoint metadata match criteria.
+    // The connection to the endpoint with metadata matching what is set in this field
+    // will use the transport socket configuration specified here.
+    // The endpoints metadata entry in *envoy.transport_socket* is used to match
+    // against the values specified in this field.
+    google.protobuf.Struct match = 2;
+
+    // The configuration of the transport socket.
+    core.TransportSocket transport_socket = 3;
+  }
+
+  // Configuration to use different transport sockets for different endpoints.
+  //
+  // The entry of *envoy.transport_socket* in endpoint :ref:`Metadata <envoy_api_msg_core.Metadata>`
+  // is used to match against the transport sockets as they appear in the list. The first
+  // :ref:`match <envoy_api_msg_Cluster.TransportSocketMatch>` is used.
+  // For example, with the following match
+  //
+  // .. code-block:: json
+  //
+  // {
+  //   "transport_socket_matches": [
+  //     {
+  //       "name": "enable-mtls",
+  //       "match": { "app": "sidecar" },
+  //       "transport_socket": { "name": "tls", "config": { ... } }
+  //     },
+  //     {
+  //       "name": "use-plaintext",
+  //       "match": {},
+  //       "transport_socket": { "name": "rawbuffer", "config": { } }
+  //     },
+  //   ]
+  // }
+  //
+  // Connection to the endpoints whose metadata value under *envoy.transport_socket*
+  // having "app"/"sidecar" key/value pair will use the "enable-mtls" socket configuration.
+  //
+  // If a :ref:`socket match <envoy_api_msg_Cluster.TransportSocketMatch>` with empty match
+  // criteria is provided, that always match any endpoint. For example, the "use-plaintext" socket
+  // configuration in case above.
+  //
+  // If an endpoint metadata's value under *envoy.transport_socket* does not match against any
+  // *TransportSockerMatch*, socket configuration fallbacks to use the *tls_context* or
+  // *transport_socket* specified in this cluster.
+  //
+  // This field allows gradual and flexible transport socket configuration changes.
+  // For example, for the same EDS typed cluster, to migrate from plain text traffic to
+  // mutual TLS traffic, management server can configure different Envoy clients with different
+  // *transport_socket_matches*.
+  // [#not-implemented-hide:]
+  repeated TransportSocketMatch transport_socket_matches = 42;
```

Comment:
nit: update `// [#comment:next free field: 42]` to 43, in the line before `message Cluster`

Reply:
Done.